include/valijson/adapters/qtjson_adapter.hpp

```diff
@@ -0,0 +1,724 @@
+     *
+     * If the referenced QtJson value is an array, this function will
+     * retrieve the number of elements in the array and store it in the output
+     * variable provided.
+     *
+     * @param   result  reference to size_t to set with result
+     *
+     * @returns true if the number of elements was retrieved, false otherwise.
+     */
+    bool getArraySize(size_t &result) const
+    {
+        if (value.isArray()) {
+            const QJsonArray array = value.toArray();
+            result = array.size();
+            return true;
+        }
+
+        return false;
+    }
+
+    bool getBool(bool &result) const
+    {
+        if (value.isBool()) {
+            result = value.toBool();
+            return true;
+        }
+
+        return false;
+    }
+
+    bool getDouble(double &result) const
+    {
+        if (value.isDouble()) {
+            result = value.toDouble();
+            return true;
+        }
+
+        return false;
+    }
+
+    bool getInteger(int64_t &result) const
+    {
+        if (value.isDouble()) {
+            result = value.toInt();
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @brief   Optionally return a QtJsonObject instance.
+     *
+     * If the referenced QtJson value is an object, this function will return a
+     * std::optional containing a QtJsonObject instance referencing the
+     * object.
+     *
+     * Otherwise it will return an empty optional.
+     */
+    opt::optional<QtJsonObject> getObjectOptional() const
+    {
+        if (value.isObject()) {
+            return opt::make_optional(QtJsonObject(value));
+        }
+
+        return opt::optional<QtJsonObject>();
+    }
+
+    /**
+     * @brief   Retrieve the number of members in the object
+     *
+     * If the referenced QtJson value is an object, this function will
+     * retrieve the number of members in the object and store it in the output
+     * variable provided.
+     *
+     * @param   result  reference to size_t to set with result
+     *
+     * @returns true if the number of members was retrieved, false otherwise.
+     */
+    bool getObjectSize(size_t &result) const
+    {
+        if (value.isObject()) {
+            const QJsonObject &object = value.toObject();
+            result = object.size();
+            return true;
+        }
+
+        return false;
+    }
+
+    bool getString(std::string &result) const
+    {
+        if (value.isString()) {
+            result = value.toString().toStdString();
+            return true;
+        }
+
+        return false;
+    }
+
+    static bool hasStrictTypes()
+    {
+        return true;
+    }
+
+    bool isArray() const
+    {
+        return value.isArray();
+    }
+
+    bool isBool() const
+    {
+        return value.isBool();
+    }
+
+    bool isDouble() const
+    {
+        return value.isDouble();
+    }
+
+    bool isInteger() const
+    {
+        return value.isDouble();
```

Comment:
I can see that QJsonValue does not make a distinction between doubles and integral values. However, we could get around this using an approach similar to the isInteger function in the json11 adapter. In that adapter, a number is considered an integer if it is possible to perform a lossless conversion to an int type.

If you were able to do that here, I would be more than happy to update the other adapter code to ensure that they are all as consistent as possible with this behaviour.